host/volume/api/http.go

```diff
@@ -110,5 +119,83 @@ func (api *HTTPAPI) Inspect(w http.ResponseWriter, r *http.Request, ps httproute
 }
 
 func (api *HTTPAPI) Snapshot(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
-	// TODO
+	volumeID := ps.ByName("volume_id")
+	snap, err := api.vman.CreateSnapshot(volumeID)
+	if err != nil {
+		switch err {
+		case volumemanager.NoSuchVolume:
+			httphelper.Error(w, httphelper.JSONError{
+				Code:    httphelper.ObjectNotFoundError,
+				Message: fmt.Sprintf("No volume by id %q", volumeID),
+			})
+			return
+		default:
+			httphelper.Error(w, err)
+			return
+		}
+	}
+
+	httphelper.JSON(w, 200, snap.Info())
+}
+
+func (api *HTTPAPI) Pull(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	volumeID := ps.ByName("volume_id")
+
+	pull := &volume.PullCoordinate{}
+	if err := json.NewDecoder(r.Body).Decode(&pull); err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+
+	hostClient, err := api.cluster.DialHost(pull.HostID)
+	if err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+
+	haves, err := api.vman.ListHaves(volumeID)
+	if err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+
+	reader, err := hostClient.SendSnapshot(pull.SnapshotID, haves)
+	if err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+
+	snap, err := api.vman.ReceiveSnapshot(volumeID, reader)
+	if err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+
+	httphelper.JSON(w, 200, snap.Info())
+}
+
+func (api *HTTPAPI) Send(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	volumeID := ps.ByName("volume_id")
+
+	haves := make([]json.RawMessage, 0)
+	if err := json.NewDecoder(r.Body).Decode(&haves); err != nil {
+		httphelper.Error(w, err)
+		return
+	}
+
+	// TODO proxy this writer...? poor error bubbling here
```

Comment:
Yeah, dropping todo.  It's not a very great story if the send command gets an error after starting the output stream, but there's not much for ways to backtrack on an http 200 that's already been flushed.